Allow publishing of packages with ignored devDependencies

georeith opened the pull request:
Resolves: https://github.com/changesets/changesets/issues/906

georeith committed:
add option to ignore devDependencies to getDependencyGraph

georeith committed:
add config to allow ignored packages in devDependencies

georeith committed:
changeset

georeith committed:
prettier

georeith committed:
add default value

georeith committed:
fix default

georeith committed:
update changeset

georeith committed:
make config optional

georeith committed:
fix config and tests

georeith committed:
fix types

georeith committed:
fix passthrough param

georeith commented:
didn't mean to close and reopen/mark this for automerge I thought I was in a different PR

kachkaev commented:
Might be worth looking at `"private": true` instead. Context: https://github.com/changesets/changesets/issues/906#issuecomment-1239468046 (**UPD2**)

Andarist commented on packages/get-dependents-graph/src/get-dependency-graph.test.ts:
```diff
@@ -115,4 +115,39 @@ describe("getting the dependency graph", function() {
     })
   );
+
+  it("should skip devDependencies if specified", function() {
```
It would also be great to have a test for the version command (I assume that it's where the original problem originated). While this package here might be directly responsible for the issue - it is an implementation detail and I would like to have a test for what people usually interact with (so `@changesets/cli`)

Andarist commented on packages/types/src/index.ts:
```diff
@@ -71,6 +71,8 @@ export type Config = {
   updateInternalDependencies: "patch" | "minor";
   ignore: ReadonlyArray<string>;
+  /** Allows packages to be published if they have ignored packages in their devDependencies */
+  allowIgnoredDevDependencies?: boolean;
```
q: what would be the downsides of removing this flag entirely and instead just ignoring such dev dependencies at all times?

wycats replied:
This would be fine, and certainly would be fine if the dev dependency is marked as private.

wycats commented:
This change would really improve my workflow 😄

Andarist commented:
@wycats I'm OK with this "feature" request - but I think that the additional config option that has been introduced here is redundant. Would you be able to prepare a PR targeting this PR that would introduce the requested changes? I'm a little bit busy right now and it might take me some days to get to this myself.

georeith committed:
Merge remote-tracking branch 'origin/main' into publish-ignore-devdeps

georeith committed:
fix tests

georeith committed:
only ignore dev deps when checking ignored packages

georeith committed:
same as previous

georeith committed:
test

georeith committed:
remove embedded yarn

georeith committed:
prettier

georeith commented:
@Andarist I reworked this removing the extra config option and made it so it just ignores devDependencies when looking at ignored files. I added a test for the version command.

I opted to not make it ignore devDependencies everywhere as devDependencies can have an effect on your files output (they may affect how it is built) but it is fine for them to be ignored.

georeith committed:
clean up test

georeith committed:
call publish from the package directory

georeith commented on packages/cli/src/commands/publish/npm-utils.ts:
```diff
@@ -193,6 +193,7 @@ async function internalPublish(
           ["publish", opts.publishDir, "--json", ...publishFlags],
           {
             env: Object.assign({}, process.env, envOverride),
+            cwd: opts.cwd,
```
Without this change my snapshot publishes fail with:

```
🦋  info Publishing "@marvelapp/crdt" at "0.0.0-20230614163206"
🦋  error an error occurred while publishing @marvelapp/crdt: E403 403 Forbidden - PUT https://registry.npmjs.org/@marvelapp%2fcrdt - You cannot publish over the previously published versions: 7.1.0.
🦋  error In most cases, you or one of your dependencies are requesting
🦋  error a package version that is forbidden by your security policy, or
🦋  error on a server you do not have access to.
🦋  error npm notice Publishing to https://registry.npmjs.org with tag dev and restricted access
🦋  error npm ERR! code E403
🦋  error npm ERR! 403 403 Forbidden - PUT https://registry.npmjs.org/@marvelapp%2fcrdt - You cannot publish over the previously published versions: 7.1.0.
🦋  error npm ERR! 403 In most cases, you or one of your dependencies are requesting
🦋  error npm ERR! 403 a package version that is forbidden by your security policy, or
🦋  error npm ERR! 403 on a server you do not have access to.
🦋  error
🦋  error npm ERR! A complete log of this run can be found in: /Users/george/.npm/_logs/2023-06-15T09_47_41_993Z-debug-0.log
🦋  error
```

georeith commented on package.json:
```diff
@@ -73,5 +73,6 @@
   },
-  "prettier": {}
+  "prettier": {},
+  "packageManager": "yarn@1.22.19"
```
For those of us who use another version of yarn than classic and have corepack enabled this will use the right yarn version when working on this repo.

Netail replied:
Bump to 1.22.22

georeith committed:
call command in publishDir

georeith committed:
fix previous

georeith committed:
remove publish change

georeith committed:
Merge branch 'main' into publish-ignore-devdeps

Netail commented:
Any update on this?

0xMurage commented:
What is preventing this from being accepted or merged?
---
